library/common/engine.cc

```diff
@@ -106,6 +107,19 @@ Engine::~Engine() {
   main_thread_.join();
 }
 
+void Engine::flushStats() {
+  // Stats must be flushed from the main thread. If we aren't on the main thread, join it.
```

Comment:
True that we need to flush from Envoy's main thread. However, I believe this won't accomplish it, as you are blocking the platform thread until Envoy's main thread finishes execution. What we need to do is:

```c++
if (server_instance_) {
  server_instance_->dispatcher().post([this]() -> void {
    server_instance_->flushStats();
  });
}
```

(sorry if there is a syntax error, you get the idea)

This way we are posting the function call into the main thread's event loop.

Reply:
Ah good call. Updating